receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go

```diff
@@ -0,0 +1,435 @@
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+)
+
+// Notes on garbage collection (gc):
+//
+// Job-level gc:
+// The Prometheus receiver will likely execute in a long running service whose lifetime may exceed
+// the lifetimes of many of the jobs that it is collecting from. In order to keep the JobsMap from
+// leaking memory for entries of no-longer existing jobs, the JobsMap needs to remove entries that
+// haven't been accessed for a long period of time.
+//
+// Timeseries-level gc:
+// Some jobs that the Prometheus receiver is collecting from may export timeseries based on metrics
+// from other jobs (e.g. cAdvisor). In order to keep the timeseriesMap from leaking memory for entries
+// of no-longer existing jobs, the timeseriesMap for each job needs to remove entries that haven't
+// been accessed for a long period of time.
+//
+// The gc strategy uses a standard mark-and-sweep approach - each time a timeseriesMap is accessed,
+// it is marked. Similarly, each time a timeseriesinfo is accessed, it is also marked.
+//
+// At the end of each JobsMap.get(), if the last time the JobsMap was gc'd exceeds the 'gcInterval',
+// the JobsMap is locked and any timeseriesMaps that are unmarked are removed from the JobsMap
+// otherwise the timeseriesMap is gc'd
+//
+// The gc for the timeseriesMap is straightforward - the map is locked and, for each timeseriesinfo
+// in the map, if it has not been marked, it is removed otherwise it is unmarked.
+//
+// Alternative Strategies
+// 1. If the job-level gc doesn't run often enough, or runs too often, a separate go routine can
+//    be spawned at JobMap creation time that gc's at periodic intervals. This approach potentially
+//    adds more contention and latency to each scrape so the current approach is used. Note that
+//    the go routine will need to be cancelled upon Shutdown().
+// 2. If the gc of each timeseriesMap during the gc of the JobsMap causes too much contention,
+//    the gc of timeseriesMaps can be moved to the end of MetricsAdjuster().AdjustMetrics(). This
+//    approach requires adding 'lastGC' Time and (potentially) a gcInterval duration to
+//    timeseriesMap so the current approach is used instead.
+
+// timeseriesinfo contains the information necessary to adjust from the initial point and to detect
+// resets.
+type timeseriesinfoPdata struct {
+	mark     bool
+	initial  *pdata.Metric
+	previous *pdata.Metric
+}
+
+// timeseriesMap maps from a timeseries instance (metric * label values) to the timeseries info for
+// the instance.
+type timeseriesMapPdata struct {
+	sync.RWMutex
+	mark   bool
+	tsiMap map[string]*timeseriesinfoPdata
+}
+
+// Get the timeseriesinfo for the timeseries associated with the metric and label values.
+func (tsm *timeseriesMapPdata) get(metric *pdata.Metric, kv pdata.AttributeMap) *timeseriesinfoPdata {
+	name := metric.Name()
+	sig := getTimeseriesSignaturePdata(name, kv)
```

Comment:
Maybe better to use a hashed object instead of string for map key

Reply:
I think the cloudwatch exporter has a pattern using a Sorted object that could work

Reply:
I'm not sure I follow.  The `pdata.AttributeMap` isn't comparable and can't be used as a map key.  Since the labels are user-provided and arbitrary I'm not sure we could create a comparable object other than a string from them.

Reply:
There's this `attribute.Distinct` type used here for something similar I think recommended by jmacd a while back

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/243d6423992a4a76db30eb254d715ccb3a23c9e4/internal/aws/metrics/metric_calculator.go#L90

I figure it has at least somewhat better performance than strings

Reply:
That is from the Go SDK's `attribute` package, which is not in use here.  These are `pdata.AttributeMap`s which would require similar conversion to utilize `attribute.Distinct`.